### PR TITLE
Cyoa

### PR DESCRIPTION
**Pull request by theundeadmonk**
The user model( login, signup, etc.) is not critical to the functionality of the app and should be treated as a black box.

### CODE REVIEW

File: cyoa/weather/README.rdoc [hunk 1]
```diff
@@ -0,0 +1,28 @@
+== README
```
Comment (djkotowski) [inline cyoa/weather/README.rdoc]: Delete unused README

File: cyoa/weather/app/assets/javascripts/account_activations.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (djkotowski) [inline cyoa/weather/app/assets/javascripts/account_activations.coffee]: Delete unused files

File: cyoa/weather/app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user,         only: [:edit, :update]
+  before_action :valid_user,       only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
+
+  def new
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/password_resets_controller.rb]: Delete these methods.  You don't need blank methods for actions.

File: cyoa/weather/app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      flash.now[:danger] = "Email address not found"
+      render 'new'
+    end
+  end
+
+  def edit
+  end
+
+  def update
+    if password_blank?
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/password_resets_controller.rb]: This looks like a model validation.

File: cyoa/weather/app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user && user.authenticate(params[:session][:password])
+      if user.activated?
+        log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
+        redirect_back_or user
+      else
+        message = "Account not activated."
+        message += "Check your email for the activation link."
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/sessions_controller.rb]: This looks like:
`Account not activated.Check your email for the activation link.`

File: cyoa/weather/app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        flash[:warning] = message
+        redirect_to root_url
+      end
+    else
+      flash.now[:danger] = 'Invalid email/password combination'
+      render 'new'
+    end
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/sessions_controller.rb]: Try to remove this arrowhead styling.  This is a large action.

File: cyoa/weather/app/controllers/static_pages_controller.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  end
+
+  def help
+  end
+
+  def about
+  end
+
+  def contact
+  end
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/static_pages_controller.rb]: Again, no need for the blank methods.  This controller can be empty, or you can use the `public` folder for static pages and remove this altogether.

File: cyoa/weather/app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  before_action :correct_user,   only: [:edit, :update]
+  before_action :admin_user,     only: :destroy
+
+  def index
+    @users = User.paginate(page: params[:page])
+  end
+
+  def show
+    @user = User.find(params[:id])
+    @climate = Climate.where date: Date.today
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/users_controller.rb]: Use parens for the argument.

File: cyoa/weather/app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      render 'new'
+    end
+  end
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
```
Comment (djkotowski) [inline cyoa/weather/app/controllers/users_controller.rb]: Since you're reusing this step, you can extract it into its own `before_action`.

File: cyoa/weather/app/helpers/application_helper.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+module ApplicationHelper
+
+  def full_title(page_title = '')
+    base_title = "Ruby on Rails Tutorial Sample App"
+    if page_title.empty?
+      base_title
+    else
+      "#{page_title} | #{base_title}"
+    end
```
Comment (djkotowski) [inline cyoa/weather/app/helpers/application_helper.rb]: L5-9 can be shrunk to two lines.
```ruby
return base_title if page_title.empty?
"#{page_title} | #{base_title}"
```

File: cyoa/weather/app/helpers/climate_helper.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    days = []
+    (0..6).each do |i|
+      if (Time.now.wday + i) > 6
+        days[i] = Date::DAYNAMES[Time.now.wday + i - 7]
+      else
+        days[i] = Date::DAYNAMES[Time.now.wday + i]
+      end
+    end
+    days
+  end
```
Comment (djkotowski) [inline cyoa/weather/app/helpers/climate_helper.rb]: This can be a `map` instead.

File: cyoa/weather/app/helpers/sessions_helper.rb [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+      if user && user.authenticated?(:remember, cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
+
+  # Returns true if the user is logged in, false otherwisee.
+  def logged_in?
+    !current_user.nil?
```
Comment (djkotowski) [inline cyoa/weather/app/helpers/sessions_helper.rb]: `current_user.present?`

File: cyoa/weather/app/helpers/sessions_helper.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    user.forget
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
+  end
+
+  # Logs out the current user.
+  def log_out
+    forget(current_user)
+    session.delete(:user_id)
+    @currrent_user = nil
```
Comment (djkotowski) [inline cyoa/weather/app/helpers/sessions_helper.rb]: Typo

File: cyoa/weather/app/models/climate.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+class Climate < ActiveRecord::Base
+
+  def self.weekly_weather
+    Weather.forecast.daily.data
+  end
+
+  def self.persist_weather
+    days = ClimateHelper.day_array
+    weekly_weather.first(7).each_with_index do |daily_weather, indx|
+      create({ max_temp:                  daily_weather.temperatureMax,
```
Comment (djkotowski) [inline cyoa/weather/app/models/climate.rb]: Alignment.

File: cyoa/weather/app/models/user.rb [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+  before_save   :downcase_email
+  before_create :create_activation_digest
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence:true, length: { maximum: 255 },
+                    format: { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, length: { minimum: 6 }, allow_blank: true
```
Comment (djkotowski) [inline cyoa/weather/app/models/user.rb]: L4-12 could be organized a bit clearer.  You've got validations, constants, and other methods (`has_secure_password`) mixed.

File: cyoa/weather/app/views/password_resets/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+      <%= hidden_field_tag :email, @user.email %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password , class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: 'form_control' %>
+
+      <%= f.submit "Update Password", class: "btn- btn-primary" %>
```
Comment (djkotowski) [inline cyoa/weather/app/views/password_resets/edit.html.erb]: Looks like a typo.  Do you want the dash at the end of the first `btn` class?

File: cyoa/weather/app/views/user_mailer/account_activation.text.erb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+Hi <%= @user.name %>
+
+Welcome to the Weather App! Click on the link below to activate your accoutn:
```
Comment (djkotowski) [inline cyoa/weather/app/views/user_mailer/account_activation.text.erb]: Typo

File: cyoa/weather/config/locales/en.yml [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+#
+#     I18n.locale = :es
+#
+# This would use the information in config/locales/es.yml.
+#
+# To learn more, please read the Rails Internationalization guide
+# available at http://guides.rubyonrails.org/i18n.html.
+
+en:
+  hello: "Hello world"
```
Comment (djkotowski) [inline cyoa/weather/config/locales/en.yml]: I don't think this translation is part of your app.

File: cyoa/weather/db/seeds.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
+# This file should contain all the record creation needed to seed the database with its default values.
```
Comment (djkotowski) [inline cyoa/weather/db/seeds.rb]: Don't use seeds for test data.

File: cyoa/weather/features/step_definitions/weather_steps.rb [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+Then(/^I should be logged in$/) do
+  expect(page.current_path).to_not have_link("Log in")
+end
+
+Given(/^I am on the sign up page$/) do
+    visit "/signup"
+end
+
+Given(/^I enter name as "(.*?)"$/) do |arg1|
```
Comment (djkotowski) [inline cyoa/weather/features/step_definitions/weather_steps.rb]: Name your args for clarity.

File: cyoa/weather/features/support/factories.rb [hunk 21]
```diff
@@ -0,0 +1,21 @@
+require 'factory_girl'
+
+FactoryGirl.define do
+  factory :user do
+    name "Cthulu"
```
Comment (djkotowski) [inline cyoa/weather/features/support/factories.rb]: Missing an 'h' in Cthulhu...

File: cyoa/weather/features/support/factories.rb [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+  factory :admin, class: User do
+    name "Herbert West"
+    email "herbert_west@thereanimator.com"
+    password "password"
+    admin true
+    activated true
+    activated_at -> { Time.zone.now }
+  end
+end
```
Comment (djkotowski) [inline cyoa/weather/features/support/factories.rb]: No Hastur?
Or Shoggoth?

File: cyoa/weather/features/user_login.feature [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+
+  Scenario: Sign up with invalid information
+    Given I am on the sign up page
+    And I enter name as ""
+    And I enter email as ""
+    And I enter password as ""
+    And I enter password confirmation as ""
+    And I click create my account
+    Then I should not be signed up
+
```
Comment (djkotowski) [inline cyoa/weather/features/user_login.feature]: Try different types of invalid information.  Not just all blank.

File: cyoa/weather/spec/models/climate_spec.rb [hunk 24]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe Climate, type: :model do
+  let(:climate) { described_class }
+
+  describe "#self.weekly_weather" do
+
+  end
```
Comment (djkotowski) [inline cyoa/weather/spec/models/climate_spec.rb]: Empty `describe`.

File: cyoa/weather/spec/models/climate_spec.rb [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+  describe "#self.persist_weather" do
+    let(:persist_weather) { described_class.persist_weather }
+
+    it "creates a Climate record" do
+      expect { persist_weather }.to change { Climate.count }.by_at_least(1)
```
Comment (djkotowski) [inline cyoa/weather/spec/models/climate_spec.rb]: Shouldn't it be `by_exactly(1)`?

File: cyoa/weather/spec/models/climate_spec.rb [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+    end
+    it "creates 7 days worth of Climate" do
+      expect { persist_weather }.to change { Climate.count }.by(7)
```
Comment (djkotowski) [inline cyoa/weather/spec/models/climate_spec.rb]: If this is here, the above test is useless.

File: cyoa/weather/test/controllers/climate_controller_test.rb [hunk 27]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ClimateControllerTest < ActionController::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (djkotowski) [inline cyoa/weather/test/controllers/climate_controller_test.rb]: Delete unused tests.

File: cyoa/weather/test/controllers/static_pages_controller_test.rb [hunk 28]
```diff
@@ -0,0 +1,28 @@
+require 'test_helper'
+
+class StaticPagesControllerTest < ActionController::TestCase
```
Comment (djkotowski) [inline cyoa/weather/test/controllers/static_pages_controller_test.rb]: No need for controller tests.  This is also very brittle, should you ever change the UI.  Just write a helper test to ensure that the title helper works correctly.

File: cyoa/weather/test/fixtures/climates.yml [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+  summary: MyString
+
+two:
+  date: 2015-04-13 12:00:40
+  max_temp: 9.99
+  min_temp: 9.99
+  precipitation_probability: 9.99
+  precipitation_type: MyString
+  humidity: 9.99
+  summary: MyString
```
Comment (djkotowski) [inline cyoa/weather/test/fixtures/climates.yml]: These two fixtures are identical.  (Also, don't use fixtures.  Use factories.)

File: cyoa/weather/test/integration/password_resets_test.rb [hunk 30]
```diff
@@ -0,0 +1,63 @@
+require 'test_helper'
+
+class PasswordResetsTest < ActionDispatch::IntegrationTest
+  def setup
+    ActionMailer::Base.deliveries.clear
+    @user = users(:michael)
+  end
+
+  test "password resets" do
```
Comment (djkotowski) [inline cyoa/weather/test/integration/password_resets_test.rb]: This is one _long_ test that is very difficult to follow.

File: cyoa/weather/test/integration/site_layout_test.rb [hunk 31]
```diff
@@ -0,0 +1,14 @@
...
+
+  test "layout links" do
+    get root_path
+    assert_template 'static_pages/home'
+    assert_select "a[href=?]", root_path, count: 2
+    assert_select "a[href=?]", help_path
+    assert_select "a[href=?]", about_path
+    assert_select "a[href=?]", contact_path
+    assert_select "a[href=?]", signup_path
+  end
```
Comment (djkotowski) [inline cyoa/weather/test/integration/site_layout_test.rb]: What benefit does this provide?

File: cyoa/weather/test/integration/user_signup_test.rb [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+    assert_not is_logged_in?
+    # Valid token, wrong email
+    get edit_account_activation_path(user.activation_token, email: 'wrong')
+    assert_not is_logged_in?
+    # Valid activation_token
+    get edit_account_activation_path(user.activation_token, email: user.email)
+    assert user.reload.activated?
+    follow_redirect!
+    assert_template 'users/show'
+    assert is_logged_in?
```
Comment (djkotowski) [inline cyoa/weather/test/integration/user_signup_test.rb]: Again, a long test that could be split up.

File: shopping_cart/features/cart_manipulation.feature [hunk 33]
```diff
@@ -0,0 +1,30 @@
+Feature: Add and remove items from cart
```
Comment (djkotowski) [inline shopping_cart/features/cart_manipulation.feature]: Remove files from other project.

**Comment by djkotowski** [comment]
Good job.  Take a look at the comments and please refactor.  The Minitest tests are hard to read and long-winded.  I highly recommend taking one or two and redoing them in Cucumber (for integration tests) or RSpec (for unit tests).  Probably one for each type.  See if you can clean and shorten the individual tests.  You also have a lot of unused files and generated comments to remove.

I don't think you need to worry about resubmitting for another review, but I would spend some time trying to improve the tests.  Consider edge cases and trim tests that have little to no value (e.g. view and controller tests).  Many of these are brittle and can report failures if you do something as harmless as change the text in your views.
